File: solutions/python/zipper/1/zipper.py
class Zipper:

    def __init__(self, value, left_node=None, right_node=None, parent_node=None):
        self._value = value
        self.left_node = left_node
        self.right_node = right_node
        self.parent_node = parent_node
    
    @staticmethod
    def from_tree(tree):

        value = tree.get("value", None)
        root = Zipper(value)
        
        left = tree.get("left", None)
        if not left is None:
            left = Zipper.from_tree(left)
            left.parent_node = root
            root.left_node = left
        
        right = tree.get("right", None)
        if not right is None:
            right = Zipper.from_tree(right)
            right.parent_node = root
            root.right_node = right

        return root

    def value(self):
        return self._value

    def set_value(self, value):
        self._value = value
        return self

    def left(self):
        return self.left_node

    def set_left(self, tree):
        new_tree = None
        if tree is not None:
            new_tree = Zipper.from_tree(tree)
            new_tree.parent_node = self
        if self.left_node is None:
            self.left_node = new_tree
            return self
        old_left_node = self.left_node
        del old_left_node
        self.left_node = new_tree
        return self

    def right(self):
        return self.right_node

    def set_right(self, tree):
        new_tree = None
        if tree is not None:
            new_tree = Zipper.from_tree(tree)
            new_tree.parent_node = self
        if self.right_node is None:
            self.right_node = new_tree
            return self
        old_right_node = self.right_node
        del old_right_node
        self.right_node = new_tree
        return self

    def up(self):
        return self.parent_node

    def to_tree(self):
        root = self
        while root.up() is not None:
            root = root.up()
        return root.to_tree_aux()


    def to_tree_aux(self):
        return {
            "value": self._value,
            "left":  None if self.left_node  is None else self.left_node.to_tree_aux(), 
            "right": None if self.right_node is None else self.right_node.to_tree_aux(),
        }


if __name__ == '__main__':

    initial = {
        "value": 1,
        "left": {
            "value": 2,
            "left": None,
            "right": {"value": 3, "left": None, "right": None},
        },
        "right": {"value": 4, "left": None, "right": None},
    }

    expected = {
        "value": 1,
        "left": {
            "value": 2,
            "left": None,
            "right": {"value": 3, "left": None, "right": None},
        },
        "right": {"value": 4, "left": None, "right": None},
    }

    zipper = Zipper.from_tree(initial)
    result = zipper.left().right().to_tree()

    print(result)

File: solutions/java/zipper/1/src/main/java/Zipper.java
import java.util.Objects;

class Zipper {
    Zipper up;
    Zipper left;
    Zipper right;

    private int value;

    public Zipper(int val) {
        this.value = val;
        this.up = null;
        this.left = null;
        this.right = null;
    }

    public BinaryTree toTree() {
        return new BinaryTree(this);
    }

    public int getValue() {
        return value;
    }

    public Zipper setLeft(Zipper leftChild) {
        this.left = leftChild;
        if (this.left == null) { return this; }
        this.left.up = this;
        return this;
    }

    public Zipper setRight(Zipper rightChild) {
        this.right = rightChild;
        if (this.right == null) { return this; }
        this.right.up = this;
        return this;
    }

    public void setValue(int val) {
        this.value = val;
    }

    @Override
    public String toString() {
        StringBuilder sbuilder = new StringBuilder();
        sbuilder.append("value: ");
        sbuilder.append(Integer.toString(this.getValue()));
        sbuilder.append(", left: ");
        if (left == null) { sbuilder.append("null"); }
        else { sbuilder.append("{ " + this.left.toString() + " }"); }
        sbuilder.append(", right: ");
        if (right == null) { sbuilder.append("null"); }
        else { sbuilder.append("{ " + this.right.toString() + " }");}
        return sbuilder.toString();
    }

}

class BinaryTree {

    private Zipper root;    

    public BinaryTree(int value) {
        this.root = new Zipper(value);
    }

    public BinaryTree(Zipper root) {
        this.root = root;
        while (this.root.up != null) {
            this.root = this.root.up;
        }
    }

    public Zipper getRoot() {
        return this.root;
    }

    public String printTree() {
        return this.root.toString();        
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || this.getClass() != o.getClass()) return false;
        BinaryTree bn = (BinaryTree)o;
        return Objects.equals(this.root, bn.getRoot());
    }

    @Override
    public int hashCode() {
        return Objects.hash(root);
    }

}
